fuse/managers/pipeline.py

```diff
@@ -0,0 +1,277 @@
+        fold sample ids.
+    :param dataset_func: Callable to a custom function that implements a dataset
+        creation. Its input is a path to cache directory and it should return a
+        train and test dataset.
+    :param train_func: Callable to a custom function that executes model training.
+    :param infer_func: Callable to a custom inference function.
+    :param eval_func: Callable to a custom evaluation function.
+    :param dataset_params: Dictionary that can contain any number of additional
+        custom parameters for dataset_func.
+    :param train_params: Dictionary that can contain any number of additional
+        custom parameters for train_func.
+    :param infer_params: Dictionary that can contain any number of additional
+        custom parameters for infer_func.
+    :param eval_params: Dictionary that can contain any number of additional
+        custom parameters for eval_func.
+    :param sample_ids_per_fold: May contain a sequence of array pairs denoting sample ids
+        for pre-defined train/validation splits. If this parameter is set to `None`,
+        the splits are decided at random.
+    """
+
+    os.environ["CUBLAS_WORKSPACE_CONFIG"] = ":4096:8"  # required for pytorch deterministic mode
+    multiprocessing.set_start_method("spawn")
+    if num_gpus_total == 0 or num_gpus_per_split == 0:
+        if train_params is not None and "manager.train_params" in train_params:
+            train_params["manager.train_params"]["device"] = "cpu"
+
+    # set debug mode:
+    mode = "default"  # Options: 'default', 'debug'. See details in FuseDebug
+    _ = FuseDebug(mode)
+
+    available_gpu_ids = FuseUtilsGPU.get_available_gpu_ids()
+    if num_gpus_total < len(available_gpu_ids):
+        available_gpu_ids = available_gpu_ids[0:num_gpus_total]
+    # group gpus into chunks of size params['common']['num_gpus_per_split']
+    gpu_resources = [
+        available_gpu_ids[i : i + num_gpus_per_split] for i in range(0, len(available_gpu_ids), num_gpus_per_split)
+    ]
+
+    # create a queue of gpu chunks (resources)
+    q_resources = Queue()
+    for r in gpu_resources:
+        q_resources.put(r)
+
+    if sample_ids_per_fold is None:
+        dataset, test_dataset = dataset_func(cache_dir=paths["cache_dir"], **dataset_params)
+        # the split decision should be the same regardless of repetition index
+        kfold = KFold(n_splits=num_folds, shuffle=True, random_state=1234)
+        sample_ids_per_fold = [item for item in kfold.split(dataset)]
+    else:
+        assert num_folds == len(sample_ids_per_fold)
+
+    for rep_index in range(num_repetitions):
+        # run training, inference and evaluation on all cross validation folds in parallel
+        # using the available gpu resources:
+        runner = partial(runner_wrapper, q_resources, rep_index, [train_wrapper, infer_wrapper, eval_wrapper])
+
+        # create process per fold
+        processes = [
```

Comment:
Do we have some kind of state (which folds completed) - so we can continue if crashed or so?

Reply:
not really, but a user could implement everything they want in their code. I assume resuming from a checkpoint should be the user responsibility.
so what exactly do you have in mind? save some text file that says "fold_X_done" when every fold is done? something else?